src/net_processing.cpp

```diff
@@ -3206,6 +3210,28 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
                 return true;
             }
         }
+        // Check for headers sync timeouts
+        if (state.fSyncStarted && state.nHeadersSyncTimeout < std::numeric_limits<int64_t>::max()) {
+            // Detect whether this is a stalling initial-headers-sync peer
+            if (pindexBestHeader->GetBlockTime() <= GetAdjustedTime() - 24*60*60) {
+                if (nNow > state.nHeadersSyncTimeout && nSyncStarted == 1 && (nPreferredDownload - state.fPreferredDownload >= 1)) {
```

Comment:
`nPreferredDownload - state.fPreferredDownload >= 1` was slightly confusing for me. It's saying we should only timeout a peer if we have another peer which is preferred. This doesn't exactly match the logic in `SendMessages()` where we'll try to start block sync from a non-preferred peer if we don't have any better options.

I think this is fine. It's not obvious that we should time out a peer if we don't have any other options which are preferred, but perhaps a comment here would make this clearer.